### PR TITLE
ARROW-11630: [Rust] Introduce limit option for sort kernel

### PR DESCRIPTION
**Pull request by sundy-li**
- Introduce `limit: Option<usize>` for sort functions, then we can use [partial_sort](https://crates.io/crates/partial_sort) to achieve better performance in queries with limit & sort. Datafusion can use it later if it can push up the limit to the sort plan (currently use None as default).

### CODE REVIEW

**Review by andygrove — COMMENTED** [review]
I took a quick look through and LGTM. I had a question about introducing the new dependency for sorting. We already have sorting in place so I'm curious why the new dependency is needed and whether we should look at updating other sorting code to use the same dependency?

**Review by jorgecarleitao — COMMENTED** [review]
I also have the same questions as @andygrove . Also, I would be more favourable having a separate function for this: sorting is already quite a complex logic already.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs:42]: Ok, I will introduce `partial_sort` with `limit` parameter.

**Comment by sundy-li** [comment]
I removed `pdqsort`, because the performance didn't show any improvement during the benches.
Added a separate function `partial_sort `as @jorgecarleitao suggested.

I did not find any other `partial_sort` function in Rust like std::partial_sort in c++.  So I created [one](https://github.com/sundy-li/partial_sort).

And the benche results about partial_sort in arrow are:
 

```
sort 2^10                     time:   [156.26 us 156.97 us 157.88 us]
sort 2^12                     time:   [753.58 us 755.43 us 758.14 us]
sort nulls 2^10               time:   [133.73 us 134.42 us 135.27 us]
sort nulls 2^12               time:   [633.41 us 635.28 us 637.51 us]

sort 2^12 limit 10            time:   [49.246 us 49.820 us 50.667 us]
sort 2^12 limit 100           time:   [115.11 us 116.26 us 117.76 us]
sort 2^12 limit 1000          time:   [113.61 us 114.21 us 114.85 us]
sort nulls 2^12 limit 10      time:   [66.283 us 66.725 us 67.347 us]
sort nulls 2^12 limit 100     time:   [66.455 us 66.932 us 67.504 us]
sort nulls 2^12 limit 1000    time:   [65.030 us 65.218 us 65.412 us]


```

File: rust/arrow/src/compute/kernels/sort.rs
Comment (paddyhoran) [inline rust/arrow/src/compute/kernels/sort.rs:660]: What does "not working good" mean here?

File: rust/arrow/src/compute/kernels/sort.rs
Comment (paddyhoran) [inline rust/arrow/src/compute/kernels/sort.rs:355]: Does `partial_sort` have a fast path when `len` is the length of the array?  The pattern in this block is repeated a few times, if we could just use `partial_sort` it could be simplified to a simple `if let` that updates `len`?

Comment (paddyhoran) [reply rust/arrow/src/compute/kernels/sort.rs:355]: Sorry not "fast path" but would the performance be the same as `sort_by`?

Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs:355]: > Sorry not "fast path" but would the performance be the same as `sort_by`?

I'm afraid not. The sort function in rust is merge sort, it is slightly faster than the heap sort for larger sets.
But it requires twice the memory of the heap sort because of the second array. 

```
sort 2^12               time:   [799.70 us 806.41 us 814.15 us]
sort 2^12 limit 2^12     time:   [1.2848 ms 1.3012 ms 1.3229 ms]

sort nulls 2^12         time:   [647.20 us 649.27 us 651.61 us]
sort nulls 2^12 limit 2^12   time:   [780.17 us 788.48 us 798.04 us]
```

We can make a new function to reduce the repeated patterns.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs]: Sorry, there is a bug in the previous version. It's fixed in the newest [commit](https://github.com/sundy-li/partial_sort/commit/372778d0011b4d647ed4ee13d1dacb0865eb266d#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R103-R141)

File: rust/arrow/src/compute/kernels/sort.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/sort.rs:1469]: I recommend using actual (values here (not just `Nan`)) as well as picking a limit other than 4 (which is the length of the array)

File: rust/arrow/src/compute/kernels/sort.rs
Comment (alamb) [inline rust/arrow/src/compute/kernels/sort.rs:1901]: here too I recommend using a limit that is less than the length of the array

File: rust/datafusion/src/physical_plan/sort.rs
Comment (alamb) [inline rust/datafusion/src/physical_plan/sort.rs:159]: The way I have seen this done in other systems is a rewrite pass a `(Sort) -> (Limit)` combination into a new operator `TopK`

**Review by alamb — COMMENTED** [review]
Thank you @sundy-li 

I found the idea of implementing "Top K"  using a sort with limit option interesting -- thank you for pointing at the clickhouse implementation. 

There are other, seemingly more memory efficient (and likely performant) ways,  to compute the `TopK`  than buffering the entire input and then doing a partial sort. For example, using a priority queue to keep only the top k values in memory. 

That being said I am not opposed to adding a `partial_sort` to arrow if others would find it useful. One datapoint is that I took a look through the C++ arrow kernels and didn't see a `limit` (which is fine):
https://github.com/apache/arrow/blob/dfd232313e1538b81a38db1e59cf4a109b61a467/cpp/src/arrow/compute/kernels/vector_sort.cc

If we choose to add this feature, another suggest I have is to inline the `partial_sort` implementation into the `arrow` crate so that:
* We can evaluate its test coverage / stability (given it has no other users yet I am not sure we can rely on the rest of the rust community to crowdsource this)
* Avoid additional crate dependencies (mostly in the future if additional code is added to `partial_sort`). 

I realize you have already gone through the trouble if creating a `partial_crate` but I think this is worth considering

File: rust/arrow/src/compute/kernels/sort.rs
Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs:1901]: Here input size is 4, expected size is 3.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/sort.rs:58]: Why does it need to be `FnMut`?

File: rust/arrow/src/compute/kernels/sort.rs
Comment (jorgecarleitao) [inline rust/arrow/src/compute/kernels/sort.rs:113]: The doc could be updated :)

**Review by jorgecarleitao — COMMENTED** [review]
I went through this PR, and also went skimmed through the code of `partial_sort`, including tests, benchmarks and the CI pipeline.

First of all, @sundy-li , thank you for your interest in this project, for this PR, and for the cool work on `partial_sort` 💯

With that said, and in summary, I am -0 on this atm.

On the one hand, there is a very good use case here as partial sort _is_ a common operation. I think it is fine that we have different kernels as C++: there is no spec about that. I also think that the API makes sense, the code is simple and easy to understand, and the tests are sufficient.

On the other hand, we are adding a 10-day old crate as a mandatory dependency of the arrow crate. This dependency is also essentially a large `unsafe` block and has no miri on CI. 

This is not to badmouth the implementation, @sundy-li, which I am fine, even though I did not go through all the details. I just feel that the sheer amount of uncommented `unsafe` on it is an indication that there is some opportunity to revisit and further improve the crate, thereby reducing the risks of undefined behavior (UB).

In summary, I wish to have this, but I sense a large risk wrt to UB.

imo, a first step here would be for `partial_sort` to mature and stabilize before we take it as a dependency. However, if others feel that the benefits already outweigh the risks, let's go for it.

**Comment by sundy-li** [comment]
> For example, using a priority queue to keep only the top k values in memory.

Yes, but lots of codes may duplicate with sort kernel. partial_sort used priority queue inside. It maybe good to do sorting in pipeline OLAP systems 

In ClickHouse,  PartialSortingTransform(Each block in each thread) --> MergeSortingTransform(Blocks to one block in each thread) --> MergingSortedTransform(N Block in N Thread to one block) . 

```
┌─explain────────────────────────────────┐
│ (Expression)                           │
│ ExpressionTransform                    │
│   (Limit)                              │
│   Limit                                │
│     (MergingSorted)                    │
│     MergingSortedTransform 16 → 1      │
│       (MergeSorting)                   │
│       MergeSortingTransform × 16       │
│         (PartialSorting)               │
│         LimitsCheckingTransform × 16   │
│           PartialSortingTransform × 16 │
│             (Expression)               │
│             ExpressionTransform × 16   │
│               (SettingQuotaAndLimits)  │
│                 (ReadFromStorage)      │
│                 NumbersMt × 16 0 → 1   │
└────────────────────────────────────────┘
```



@alamb @jorgecarleitao  Thanks for all your reviews. I also have the consideration about unsafe codes in partial_sort may break `arrow`, because it was just created, without any used in production(BTW I am new to rust).

We can keep this MR open currently until you think it's safe enough or must have it.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs:58]: Just to be compatible with `std::sort_by`.

**Comment by Dandandan** [comment]
I think there are some very nice use-cases of this optimization (top-k kind of queries), I think it would be very cool to have more advanced features like this in Arrow/DataFusion/...

The partial_sort crate seems also quite small (about 60 lines without tests/imports). What about having the code in Arrow for now, and (try to) limit the amount of unsafety @sundy-li and we can have a next review iteration?

**Comment by alamb** [comment]
> In ClickHouse, PartialSortingTransform(Each block in each thread) 

This is a very cool idea (effectively to compute top-k for each block in parallel and then merge them together). Thank you for the explanation @sundy-li

**Comment by jorgecarleitao** [comment]
Thanks @sundy-li for the explanation.

> The partial_sort crate seems also quite small (about 60 lines without tests/imports). What about having the code in Arrow for now, and (try to) limit the amount of unsafety @sundy-li and we can have a next review iteration?

I do not follow the argument. Isn't the whole purpose of crates and dependencies to separate concerns and allow code reuse across the ecosystem without dragging dependencies?

From what I've seen, the crate:

* addresses one problem and one alone (unlike our repo or our workspace)
* has CI and CD in place
* it is easy to publish in crates.io (via a github action) which makes CD also easy
* it is duo license, as it is the standard in Rust
* it uses SemVer

If I were @sundy-li, I would keep the setup exactly like it is: it enables the implementation to be used by more projects without having to drag arrow and its dependencies with it, while at the same time maintains the generality that the algorithm itself has. I would be behaving exactly how @sundy-li is behaving: engage with potential "customers" like Arrow to identify value, how it can be adopted, and what are the points to improve.

From our side, I am not sure I agree with copy-pasting code from a crate to our own crate. IMO we should instead actively engage with our "providers" by contributing upstream. In this particular case, one step is to have MIRI run on CI. The other is to either limit or document the usage of `unsafe`. This way, the Rust community as a whole benefits from these, instead of the benefit be limited to users of the Arrow crate. I created a PR upstream to address one of the steps: https://github.com/sundy-li/partial_sort/pull/1 .

Side note: @alamb , what just happened (on which I could trivially PR upstream without any complications, issues, etc.) is a concrete example of what I have been trying to express on the mailing list wrt to the UX that IMO our repo lacks: I can't do that on this repo.

**Comment by Dandandan** [comment]
@jorgecarleitao 

I'm perfectly fine with that approach too.

I do think that depending on crates that are useful instead of building our own or even duplicating code is what we should do in general, and indeed will benefit others in the Rust community. 

My reasoning behind my suggestion above was

* The crate has no currently no users and (I think) only 1 maintainer.
* The code size is small
* The PR author is the same as the maintainer of the crate (and I guess it was created for this use case for now?)
* It would be easier to review the code / run Miri checks, etc. when it was included in the PR

Of course, there might be some other parties that would be interested in `partial_sort` or maybe it will at some point make it to the standard library.

Also dependencies have *some* costs, some of them being:

* Has some risk that the crated will become unmaintained at some point, but causes a bug / security problem, etc. in Arrow. Or we would like to improve it, but a crate maintainer might be unresponsive. A recent example of this I saw in the `memmap` crate which now has a fork `memmap2` which is quite hard to find as the docs / repo of `memmap` are still online.
* Non-semver breaking changes can not be caught in PRs
* Build times become larger compared to having the same code in the project (as the crate will be downloaded / compiled separately). Of course one extra small dependency won't matter that much, but it accumulates with lots of dependencies.

**Comment by jorgecarleitao** [comment]
Thanks for that, @Dandandan, all great points  👍, to which I generally agree with.

> It would be easier to review the code / run Miri checks, etc. when it was included in the PR

I agree. My concern atm is that our own crate does not pass miri checks (our miri checks are `cargo miri test || true`), thus, any potential soundness issues new code may have cannot be identified by our own CI. I only recently was able to have [a version of a arrow crate](https://github.com/jorgecarleitao/arrow2) to pass miri checks, and I am still [deactivating some checks](https://github.com/jorgecarleitao/arrow2/blob/main/.github/workflows/test.yml#L160) and investigating. My hypothesis is that it is very difficult to write `safe` code in this crate's design and thus my state of mind is that our best bet is to push the soundness checks about specific parts of the code to places where those can be done in a more contained environment (i.e. without all the transmutes and the like that we do here). We should still of course try to fix our own miri checks. xD

> Build times become larger compared to having the same code in the project (as the crate will be downloaded / compiled separately). Of course one extra small dependency won't matter that much, but it accumulates with lots of dependencies.

I agree. And I think that this is already a problem for DataFusion and parquet. I do get the feeling that the culprits there are crossbeam, tokio, tower and the like. These small crates seem to be downloaded and compiled in parallel and thus have a small impact when compared to the big beasts that block compilation (i.e. we have unbalanced-sized tasks). In arrows' case, the largest is flatbuffer I think, which we could feature-gate under e.g. `io_ipc` in case people do not need that. We can't get away with these during testing, but I think it is by design; if we want, IMO we need a new compilation unit (e.g. a crate).

> Has some risk that the crated will become unmaintained at some point, but causes a bug / security problem, etc. in Arrow. Or we would like to improve it, but a crate maintainer might be unresponsive. A recent example of this I saw in the memmap crate which now has a fork memmap2 which is quite hard to find as the docs / repo of memmap are still online.

I agree. That IMO should be sufficient for us to bring the crate to arrow.

**Comment by alamb** [comment]
> From our side, I am not sure I agree with copy-pasting code from a crate to our own crate. IMO we should instead actively engage with our "providers" by contributing upstream.

@jorgecarleitao as you and @Dandandan have elucidated,  I think there are tradeoffs to the two approaches. I am happy with either approach in general -- I suggested the "include it in arrow approach" in this case as I felt it had less overhead for @sundy-li  who has already worked on this PR a great deal, but that may not be the best approach either short or long term.

**Comment by sundy-li** [comment]
Hi, all.  From other user's adivice, partial_sort could be 
```
pub fn partial_sort<T, F>(v: &mut [T], limit: usize, mut is_less: F)
where
    F: FnMut(&T, &T) -> Ordering,
{
    let (before, _mid, _after) = v.select_nth_unstable_by(limit, &mut is_less);
    before.sort_unstable_by(is_less);
}
```

So I abandon to use of custom partial_sort. This could fix the issue and risk we are arguing about.

**Review by alamb — APPROVED** [review]
I think this PR is looking good now. Given the lengthy discussion I would like at least one other person (e.g @jorgecarleitao  @paddyhoran @andygrove ) to give their signoff on this PR as well 

Thank you for sticking with it @sundy-li  and I really like how this PR has shaped up.

File: rust/arrow/src/compute/kernels/sort.rs
Comment (Dandandan) [inline rust/arrow/src/compute/kernels/sort.rs:838]: I am wondering, why can we use unstable sort here, while we use stable sort for the normal sort kernel?
Shouldn't that be an option of the kernel?

Comment (sundy-li) [reply rust/arrow/src/compute/kernels/sort.rs:838]: There are some reasons for using unstable_sort.  
1. `select_nth_unstable_by` is unstable, so using `before.sort_stable_by`cadon't ensure it's stable.
2. `sort_unstable` is faster than sort_stable, refer to [doc](https://doc.rust-lang.org/std/primitive.slice.html#method.sort_unstable)

Comment (Dandandan) [reply rust/arrow/src/compute/kernels/sort.rs:838]: Clear, thanks @sundy-li.

Maybe some docs can be added about this?

I think it probably is not really a problem, but might be surprising by someone using the kernel. And we might want to see whether we could use a (faster) unstable version of the normal sort kernel as well (I think that could be used by DataFusion as well).

What do you think @alamb

**Comment by alamb** [comment]
Any last thoughts prior to merge? I'll plan to merge this later today unless I hear otherwise

**Comment by alamb** [comment]
Merged. Thanks again @sundy-li  for getting this done. It is a cool piece of functioanlity

**Comment by alamb** [comment]
BTW when I was writing some doc examples for this functionality, it occured to me that there is another potential way to expose this functionality -- https://github.com/apache/arrow/pull/9722

**Pull request closed without merging**
